File: backend/src/routes/featured.rs
use axum::{Json, Router, extract::State, routing::get};
use tracing::instrument;

use crate::{error::AppResult, models::FeaturedPair, repo::FeaturedRepo, state::AppState};

const DEFAULT_PAIR_COUNT: usize = 6;

pub fn router() -> Router<AppState> {
    Router::new().route("/", get(list_featured_pairs))
}

/// GET /api/featured — Random popular shader pairs for the homepage hero
#[instrument(skip(state))]
async fn list_featured_pairs(State(state): State<AppState>) -> AppResult<Json<Vec<FeaturedPair>>> {
    let pairs = FeaturedRepo::random_pairs(state.db(), DEFAULT_PAIR_COUNT).await?;
    Ok(Json(pairs))
}

File: frontend/src/lib/api/endpoints/featured.ts
import type { FeaturedPair } from '$lib/bindings';
import type { Result } from 'true-myth';
import { ApiClient } from '../client';
import type { ApiError } from '../errors';

export class FeaturedEndpoints extends ApiClient {
	/**
	 * Get random popular shader pairs for the homepage hero
	 */
	list(): Promise<Result<FeaturedPair[], ApiError>> {
		return this.get<FeaturedPair[]>('/api/featured');
	}
}
